### PR TITLE
MASS objective + transform

### LINKED ISSUE #15: Implement MASS objective for denoising autoencoder
According to Baziotis et al [1], the MASS objective [2] is superior to BART, for training on monolingual data using the denoising sequence auto-encoder approach.

BART (alreday implemented in mammoth) is easier to implement though, as it fits with the transform framework. MASS on the other hand requires:

- [x] Two separate target language sequences: decoder input and output. (x_5 is replaced by a mask symbol in the input)
- [ ] Token-level masking of the loss. Only the three unmasked tokens x_3--x_5 contribute to the loss. The other tokens can be generated (as it is easier than not doing so), but their contribution to the loss should be zeroed out.

Both MASS and BART use sampled noise, and must be applied online during training, as opposed to cleaning preprocessing (#13) that can also be applied offline.

Note that MASS is tricky to implement, while BART we already have. It may not be worth the effort to implement this.

image

[1] Baziotis et. al (2023) "When Does Monolingual Data Help Multilingual Translation: The Role of Domain and Model Scale" https://arxiv.org/abs/2305.14124
[2] Song et. al (2019) "MASS: Masked Sequence to Sequence Pre-training for Language Generation" https://arxiv.org/abs/1905.02450

### PR DESCRIPTION
**Pull request by TimotheeMickus**
closes #15 

Essentially, adds a `labels` key to batches to distinguish between decoder-side inputs (`tgt` and `targets'`) and decoder-side labels (currently only in use for MASS, but can be helpful for other decoder-side denoising schemes).
`labels` are initialized as a copy of `target`  unless otherwise specified by a transform.

With this change, the MASS noising transform devolves into a slightly more complex formatting in a BART noising transform.

### CODE REVIEW

**Comment by TimotheeMickus** [comment]
smoketested on unpc. requires a thorough review.

File: onmt/transforms/denoising.py [hunk 1]
```diff
@@ -322,10 +322,11 @@ def __repr__(self):
         return '{}({})'.format(cls_name, cls_args)
 
 
-@register_transform(name='bart')
-class BARTNoiseTransform(Transform):
+@register_transform(name='ae_noise')
+class NoiseTransform(Transform):
     def __init__(self, opts):
         super().__init__(opts)
+        self.denoising_objective = opts.denoising_objective
```
Comment (Waino) [inline onmt/transforms/denoising.py:329]: We should validate that `mask-random` is not used with the MASS objective.

In BART, it is possible to occasionally substitute a random token instead of the mask token. This slightly alleviates the tendency to copy all unmasked tokens verbatim.

In MASS, substituting a random token on the source side would also affect the target sequence: the random token would be complemented into a mask token, and would not contribute to the loss. This does not make sense.

**Review by Waino — CHANGES_REQUESTED** [review]
It looks like this should work. Some changes would improve clarity.

File: onmt/utils/loss.py [hunk 2]
```diff
@@ -328,6 +330,7 @@ def _make_shard_state(self, batch, output, range_, attns=None):
         shard_state = {
             "output": output,
             "target": batch.tgt[range_start:range_end, :, 0],
+            "labels": batch.labels[range_start:range_end, :, 0],
```
[the remark this answers is not in the document]
Comment (TimotheeMickus) [reply onmt/utils/loss.py:333]: I'm strongly in favor of the currently implemented approach here (I'd rather anything that exits the batchers to have roughly the same attributes)

File: onmt/trainer.py [hunk 3]
```diff
@@ -461,7 +461,7 @@ def _gradient_accumulation_over_lang_pairs(
             seen_comm_batches.add(comm_batch)
             if self.norm_method == "tokens":
                 num_tokens = (
-                    batch.tgt[1:, :, 0].ne(self.train_loss_md[f'trainloss{metadata.tgt_lang}'].padding_idx).sum()
+                    batch.labels[1:, :, 0].ne(self.train_loss_md[f'trainloss{metadata.tgt_lang}'].padding_idx).sum()
                 )
                 normalization += num_tokens.item()
             else:
```
[the remark this answers is not in the document]
Comment (TimotheeMickus) [reply onmt/trainer.py:467]: Yes.

Full disclosure: I did minimal edits. I don't know why the code doesn't use `batch.tgt` directly line 490. I try not to fix what's not broken.

File: onmt/transforms/denoising.py (hunk 1 above)
Comment (TimotheeMickus) [reply onmt/transforms/denoising.py:331]: so `random_ratio > 0`

File: onmt/utils/loss.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (Waino) [reply onmt/utils/loss.py:334]: Yes, consistent naming and structure is desirable, and is a good enough motivation for keeping target here.

On the topic of consistent naming: I noticed that there are some places (at least in the `_stats` method) where the name `target` is used, but it now actually refers to `labels`. `_stats` takes `gtruth`, which was changed in this PR.

**Review by Waino — APPROVED** [review]
LGTM
